src/Featurizers/Base/RollingWindowTransformerBase.h

```diff
@@ -0,0 +1,187 @@
+///
+template <
+    typename InputT,
+    typename OutputT,
+    size_t MaxNumTrainingItemsV
+>
+class RollingWindowTransformerBase:
+    public Components::InferenceOnlyTransformerImpl<InputT, std::vector<OutputT>> {
+public:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Types
+    // |
+    // ----------------------------------------------------------------------
+    using BaseType = Components::InferenceOnlyTransformerImpl<InputT, std::vector<OutputT>>;
+    using CalculatorFunctionInputType = typename Components::CircularBuffer<InputT>::iterator;
+    using CalculatorFunction = std::function<OutputT(CalculatorFunctionInputType, CalculatorFunctionInputType)>;
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Methods
+    // |
+    // ----------------------------------------------------------------------
+    RollingWindowTransformerBase(std::uint32_t maxWindowSize, CalculatorFunction calculator, std::uint32_t horizon, std::uint32_t minWindowSize) ;
+    ~RollingWindowTransformerBase(void) override = default;
+
+    FEATURIZER_MOVE_CONSTRUCTOR_ONLY(RollingWindowTransformerBase);
+
+    bool operator==(RollingWindowTransformerBase const &other) const;
+    bool operator!=(RollingWindowTransformerBase const &other) const;
+
+protected:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Protected Members
+    // |
+    // ----------------------------------------------------------------------
+    // These are protected so that the derived classes can use them for saving of state.
+    const std::uint32_t                             _maxWindowSize;
+    const std::uint32_t                             _minWindowSize;
+    const std::uint32_t                             _horizon;
+
+private:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Private Members
+    // |
+    // ----------------------------------------------------------------------
+    CalculatorFunction                              _calculator;
+    Components::CircularBuffer<InputT>              _buffer;
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Methods
+    // |
+    // ----------------------------------------------------------------------
+
+    // MSVC runs into problems when the declaration and definition are separated
+    void execute_impl(typename BaseType::InputType const &input, typename BaseType::CallbackFunction const &callback) override {
+        _buffer.push(input);
+        const size_t bufferSize = _buffer.size();
+
+        typename BaseType::TransformedType         results(_horizon);
+
+        for (std::uint32_t offset = 0; offset < _horizon; ++offset){
+            OutputT result;
+            size_t numElements;
+            size_t startingOffset;
+

```

Comment:
These values are only used in the else clause - define them there. As a general rule, define variables as close as possible to where they are used. Bonus points if you can declare the vars as const.

Reply:
I moved them into the else clause, but since they have to be declared before they are initialized I can't make them const.